### PR TITLE
ClangFormat the C code

### LINKED ISSUE #4493: ClangFormat the C code
Run [ClangFormat](https://clang.llvm.org/docs/ClangFormat.html) on all of the C code to once and for all get rid of the mixed 8 tab/spaces -> all 4 spaces, get all the braces on a consistent style, enclose single lines in braces, and generally make it all consistent.

Re: https://github.com/python-pillow/Pillow/pull/1770#issuecomment-196974405

Will prevent bugs like https://github.com/python-pillow/Pillow/pull/4492.

Let's do this after the 2020-04-01 release (https://github.com/python-pillow/Pillow/issues/4354), when the PR list will be at a local minimum and to avoid merge conflicts.

There's four [pre-commit hooks](https://pre-commit.com/hooks.html) available, let's pick a good one to add to `.pre-commit-config.yaml` so the CI will prevent regressions.

We'll need to pick a [style](https://clang.llvm.org/docs/ClangFormatStyleOptions.html) or define our own. Is there a predefined one close to the current style, or that is commonly used for Python extensions? What do other Python projects use?

### PR DESCRIPTION
**Pull request by hugovk**
Fixes #4493. 

I found a "clang-format style that approximates Python's [PEP 7](https://www.python.org/dev/peps/pep-0007/)":

* https://github.com/pganssle/zoneinfo/blob/master/.clang-format

(The zoneinfo repo was the reference implementation for `zoneinfo` in the stdlib (PEP 615), and is now a backport for Python 3.6+.)

This style has also been used elsewhere, as is or as the basis:

* https://github.com/search?q=%22A+clang-format+style+that+approximates+Python%27s+PEP+7%22&type=Code

Including CPython's new PEG generator (written by Guido and two others), which is the same as the zoneinfo one but with `ColumnLimit` increased from 79 to 95:

* https://github.com/python/cpython/blob/master/Tools/peg_generator/.clang-format

Here's how it looks applied to Pillow:

```sh
clang-format -i `git ls-tree -r HEAD --name-only | grep "\.[ch]$"`
```

1. The first commit is the same config file as in zoneinfo
2. The second formats the codebase with the config
3. The third increases `ColumnLimit` from 79 to 88, to match Black
4. The fourth formats again with the 88 limit

I suggest we use this, and welcome feedback and any further tweaks.

Before merge, to make the history cleaner, I'll squash the config to a single commit, and the formatting to another single commit.

Also this will need adding to the CI.

### CODE REVIEW

**Comment by wiredfool** [comment]
I really dislike this form of function definition, especially since almost all other forms of braces are inline with the end of the starting line:

```
static PyObject*
_putdata(ImagingObject* self, PyObject* args)
{
```

The rest looks pretty reasonable. I think this is forcing braces around single line blocks (if/while), which is good. If it's not, then I'd say that it should.

**Comment by nulano** [comment]
I think this ignores the following part of PEP 7:
> * Breaking long lines: if you can, break after commas in the outermost argument list. Always indent continuation lines appropriately, e.g.:
> ```c
> PyErr_Format(PyExc_TypeError,
>             "cannot create '%.100s' instances",
>             type->tp_name);
> ```

This is also a requirement in Python code in PEP 8 and Black.

**Comment by hugovk** [comment]
> I really dislike this form of function definition, especially since almost all other forms of braces are inline with the end of the starting line:
> 
> ```c
> static PyObject*
> _putdata(ImagingObject* self, PyObject* args)
> {
> ```

This config sets `BreakBeforeBraces: Stroustrup`, overriding the Google value of `BreakBeforeBraces: Attach`.

The options ([docs with examples](https://clang.llvm.org/docs/ClangFormatStyleOptions.html)):

* Attach: Always attach braces to surrounding context.
* Linux: Like Attach, but break before braces on function, namespace and class definitions.
* Stroustrup: Like Attach, but break before function definitions.
* Allman: Always break before braces.
* GNU: Always break before braces and add an extra level of indentation to braces of control statements, not to those of class, function or other definitions.

Attach is the only one not to break before that brace:

```c
static PyObject *
_putdata(ImagingObject *self, PyObject *args) {
```

I've changed config to `BreakBeforeBraces: Attach` in ba2242157529e29fef96fbcad4d32c162876e6b9 and formatted with it in 4466b71e26c7db1780f6366c6ac5e8c208578498. Is it better?

**Comment by hugovk** [comment]
> I think this is forcing braces around single line blocks (if/while), which is good. If it's not, then I'd say that it should.

I was hoping it would, to prevent bugs like https://github.com/python-pillow/Pillow/pull/4492, but alas it's out of scope for clang-format (https://stackoverflow.com/a/36871872/724176).

@radarhere fixed up the existing code in https://github.com/python-pillow/Pillow/pull/4618.

**Comment by hugovk** [comment]
> I think this ignores the following part of PEP 7:
> 
> > * Breaking long lines: if you can, break after commas in the outermost argument list. Always indent continuation lines appropriately, e.g.:
> > 
> > ```c
> > PyErr_Format(PyExc_TypeError,
> >             "cannot create '%.100s' instances",
> >             type->tp_name);
> > ```
> 
> I think this is also a requirement in Python code in Black.

Looks like this can be achieved with (https://stackoverflow.com/a/47097008/724176):

```
AlignAfterOpenBracket: AlwaysBreak
BinPackArguments: false
BinPackParameters: false
```

Docs: https://clang.llvm.org/docs/ClangFormatStyleOptions.html
Config: f714356374b84fe6018b7ad4af576c45c0fbc7a5
Formatted: a48cc59e29433052411dab4482c99639d8d7379f

**Comment by hugovk** [comment]
I've rebased, squashed the style modifications to a single commit, and reformatted.

File: src/_imaging.c [hunk 1]
```diff
@@ -94,40 +94,38 @@
...
+#define S16(v) ((v) < 32768 ? (v) : ((v)-65536))
 
 /* -------------------------------------------------------------------- */
 /* OBJECT ADMINISTRATION                        */
 /* -------------------------------------------------------------------- */
 
 typedef struct {
-    PyObject_HEAD
-    Imaging image;
+    PyObject_HEAD Imaging image;
```
Comment (nulano) [inline src/_imaging.c:120]: This doesn't look right. Perhaps the following should get formatted properly?
```suggestion
    PyObject_HEAD;
    Imaging image;
```
This change can be done in all files with a simple grep.

**Review by nulano — COMMENTED** [review]
> Let's do this after the 2020-04-01 release (#4354), when the PR list will be at a local minimum and to avoid merge conflicts.

Maybe after 8.0.0 (#4764)?

File: src/_imaging.c (hunk 1 above)
Comment (radarhere) [reply src/_imaging.c:120]: Adding in the semicolon causes syntax errors on Windows (AppVeyor and GHA).

Comment (nulano) [reply src/_imaging.c:120]: Ah great, leave it to MSVC to apply arbitrary restrictions from a standard it doesn't even fully support yet.

After a bit of effort, I have found the following hack which works with both MSVC and GCC 4.1 and has no effect on structure size (see https://godbolt.org/z/76W7f4):
```suggestion
    PyObject_HEAD int:0;
    Imaging image;
```
The extra code can be hidden in a define such as `#define PILLOW_HEAD PyObject_HEAD int:0` and used similarly to my original comment. However, I'm not sure whether it is worth the extra effort.

**Comment by radarhere** [comment]
I don't have strong feelings about this.
We're coming up on a release. Is it a good time to merge?

File: src/_imaging.c (hunk 1 above)
Comment (wiredfool) [reply src/_imaging.c:120]: I think we should leave this if possible, PyObject_HEAD is the standard way to do it, and hiding it in a macro is additional complication that doesn’t make sense except for formatting.

**Comment by wiredfool** [comment]
Let’s merge after release. Fewer potential merge errors and rebase required then.

**Pull request merged**